src/main/java/com/google/devtools/build/remote/client/LogParserUtils.java

```diff
@@ -61,14 +67,104 @@ private FileInputStream openGrpcFileInputStream() throws ParamException, IOExcep
     return new FileInputStream(filename);
   }
 
+  // Returns an ExecuteResponse contained in the operation and null if none.
+  // If the operation contains an error, returns null and populates StringBuilder "error" with
+  // the error message.
+  public static <T extends Message> T getExecuteResponse(
+      Operation o, Class<T> t, StringBuilder error) throws IOException {
+    if (o.getResultCase() == ResultCase.ERROR && o.getError().getCode() != Code.OK.value()) {
+      error.append(o.getError().toString());
+      return null;
+    }
+    if (o.getResultCase() == ResultCase.RESPONSE && o.getDone()) {
+      return o.getResponse().unpack(t);
+    }
+    return null;
+  }
+
+  // Returns a Digest contained in the log entry, and null if none.
+  public static Digest extractDigest(LogEntry entry) {
+    if (!entry.hasDetails()) {
+      return null;
+    }
+    RpcCallDetails details = entry.getDetails();
+    if (details.hasExecute()) {
+      if (details.getExecute().hasRequest()
+          && details.getExecute().getRequest().hasActionDigest()) {
+        return details.getExecute().getRequest().getActionDigest();
+      }
+    }
+    if (details.hasGetActionResult()) {
+      if (details.getGetActionResult().hasRequest()
+          && details.getGetActionResult().getRequest().hasActionDigest()) {
+        return details.getGetActionResult().getRequest().getActionDigest();
+      }
+    }
+    return null;
+  }
+
+  private static List<ExecuteResponse> extractExecuteResponse(List<Operation> operations)
+      throws IOException {
+    ArrayList<ExecuteResponse> result = new ArrayList<>();
+    for (Operation o : operations) {
+      StringBuilder error = new StringBuilder();
+      ExecuteResponse response = LogParserUtils.getExecuteResponse(o, ExecuteResponse.class, error);
+      if (response != null
+          && (response.hasResult()
+              || (response.hasStatus()) && response.getStatus().getCode() != Code.OK.value())) {
+        result.add(response);
+      }
+    }
+    return result;
+  }
+
+  // Returns a list of ExecuteResponse messages contained in the log entry,
+  // an empty list if there are none.
+  // If the LogEntry contains a successful cache lookup, an ExecuteResponse is constructed
+  // with that ActionResult and cached_result set to true.
+  public static List<ExecuteResponse> extractExecuteResponse(LogEntry entry) throws IOException {
```

Comment:
I think these functions will be a lot simpler if you always return a single response, and fail as soon as you get more than one (rather than returning a whole list and checking at the very top that it contains one element). 

Getting more than one is an error, this is explicitly against the API definition, and would be a server bug. To me, this is a case where raising an exception (for example with Precondition.fail) makes sense.

Reply:
I've modified the error conditions in ActionGrouping:
- give an error on multiple ExecutionResponse for a single LogEntry.
- allow multiple ExecutionreResponses across LogEntries, as long as the ones that get overridden don't have Result (should not error for valid retries).

I've tried replacing List<ExecutionResponse> with a single result, but I didn't like how it looked.
The code doesn't become much simpler. But now the error message construction is moved from ActionGrouping to LogParserUtils. 

The problem is, LogParserUtils doesn't have enough information to give a useful error message (e.g., which action was this?). That would mean that ActionGrouping would have to catch the exception to add information to the error message.

WDYT?

Reply:
Ah, right. Never mind, then. It will all probably look very different in Go anyway :-)